### DISCUSSION TITLE
[config]solution to add additional time grain TIME_GRAIN_ADDONS

### DISCUSSION THREAD

**Discussion by iercan** [apache/superset#13923] [Q&A / Help] [answered]
I'm trying to add extra time grains. I added this to the config
```
TIME_GRAIN_ADDONS = {
        'PT2S': '2 second',
        'PT5S': '5 second',
        'PT10S': '10 second',
        'PT30S': '30 second'
        }
```
I run `superset init` and restarted superset but I can not see them in ui. Maybe there is something I am missing?

### Environment

- superset version: 1.0.1 installed va docker
- python version: 3.7
- database: druid 20.1

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

**Comment by villebro** [comment]
You also need to implement the added time grains per database engine. Check the `TIME_GRAIN_ADDON_EXPRESSIONS` config parameter. Closing for now, please reopen if this doesn't work.

**Comment by iercan** [comment, marked as answer]
Adding it solved problem. thanks. I'm putting here in case someone needs

```
TIME_GRAIN_ADDONS: Dict[str, str] = {
        'PT2S': '2 second',
        'PT5S': '5 second',
        'PT10S': '10 second',
        'PT30S': '30 second'
        }

TIME_GRAIN_ADDON_EXPRESSIONS: Dict[str, Dict[str, str]] = {
        'druid': {
            'PT2S': '''TIME_FLOOR({col}, 'PT2S')''',
            'PT5S': '''TIME_FLOOR({col},  'PT5S')''',
            'PT10S': '''TIME_FLOOR({col}, 'PT10S')''',
            'PT30S': '''TIME_FLOOR({col}, 'PT30S')'''
            }
        }
```

**Comment by riceliu** [comment]
> Adding it solved problem. thanks. I'm putting here in case someone needs
> 
> ```
> TIME_GRAIN_ADDONS: Dict[str, str] = {
>         'PT2S': '2 second',
>         'PT5S': '5 second',
>         'PT10S': '10 second',
>         'PT30S': '30 second'
>         }
> 
> TIME_GRAIN_ADDON_EXPRESSIONS: Dict[str, Dict[str, str]] = {
>         'druid': {
>             'PT2S': '''TIME_FLOOR("__time", 'PT2S')''',
>             'PT5S': '''TIME_FLOOR("__time",  'PT5S')''',
>             'PT10S': '''TIME_FLOOR("__time", 'PT10S')''',
>             'PT30S': '''TIME_FLOOR("__time", 'PT30S')'''
>             }
>         }
> ```

Can you help give me the expression function of mysql?

**Comment by iercan** [comment]
@riceliu try this

```
TIME_GRAIN_ADDONS: Dict[str, str] = {
        'PT2S': '2 second',
        'PT5S': '5 second',
        'PT10S': '10 second',
        'PT30S': '30 second'
        }

TIME_GRAIN_ADDON_EXPRESSIONS: Dict[str, Dict[str, str]] = {
        'mysql': {
            'PT2S': '''FROM_UNIXTIME(UNIX_TIMESTAMP({col})-MOD(UNIX_TIMESTAMP({col}), 2))''',
            'PT5S': '''FROM_UNIXTIME(UNIX_TIMESTAMP({col})-MOD(UNIX_TIMESTAMP({col}), 5))''',
            'PT10S': '''FROM_UNIXTIME(UNIX_TIMESTAMP({col})-MOD(UNIX_TIMESTAMP({col}), 10))''',
            'PT30S': '''FROM_UNIXTIME(UNIX_TIMESTAMP({col})-MOD(UNIX_TIMESTAMP({col}), 30))'''
            }
        }
```
